test/ShallowWrapper-spec.js

```diff
@@ -2455,6 +2455,575 @@ describe('shallow', () => {
+            [
+              'componentWillReceiveProps',
+              { foo: 'bar' }, { foo: 'baz' },
+              { foo: 'context' },
+            ],
+            [
+              'componentWillUpdate',
+              { foo: 'bar' }, { foo: 'baz' },
+              { foo: 'state' }, { foo: 'state' },
+              { foo: 'context' },
+            ],
+            [
+              'componentDidUpdate',
+              { foo: 'bar' }, { foo: 'baz' },
+              { foo: 'state' }, { foo: 'state' },
+              { foo: 'context' },
+            ],
+          ]
+        );
+      });
+
+      it('should cancel rendering when Component returns false in shouldComponentUpdate', () => {
+        const spy = sinon.spy();
+
+        class Foo extends React.Component {
+          shouldComponentUpdate() {
+            return false;
+          }
+          componentWillUpdate() {
+            spy();
+          }
+          componentDidUpdate() {
+            spy();
+          }
+          render() {
+            return <div>foo</div>;
+          }
+        }
+
+        const wrapper = shallow(<Foo foo="bar" />, { lifecycleExperimental: true });
+        wrapper.setProps({ foo: 'baz' });
+        expect(spy.callCount).to.equal(0);
+      });
+
+      it('should be batching updates in componentWillReceiveProps', () => {
+        const spy = sinon.spy();
+        class Foo extends React.Component {
+          constructor(props) {
+            super(props);
+            this.state = {
+              count: 0,
+            };
+          }
+          componentWillReceiveProps() {
+            this.setState({ count: this.state.count + 1 });
+            this.setState({ count: this.state.count + 1 });
+          }
+          render() {
+            spy();
+            return <div>{this.props.foo}</div>;
+          }
+        }
+        const result = shallow(<Foo />, { lifecycleExperimental: true });
+        result.setProps({ name: 'bar' });
+        expect(result.state('count')).to.equal(1);
+        expect(spy.callCount).to.equal(2);
+      });
+
+      it('should be batching updates in componentWillUpdate', () => {
+        const spy = sinon.spy();
+        class Foo extends React.Component {
+          constructor(props) {
+            super(props);
+            this.updated = false;
+            this.state = {
+              count: 0,
+            };
+          }
+          componentWillUpdate() {
+            if (!this.updated) {
+              this.updated = true;
+              this.setState({ count: this.state.count + 1 });
+              this.setState({ count: this.state.count + 1 });
+            }
+          }
+          render() {
+            spy();
+            return <div>{this.props.foo}</div>;
+          }
+        }
+        const result = shallow(<Foo />, { lifecycleExperimental: true });
+        result.setProps({ name: 'bar' });
+        expect(result.state('count')).to.equal(1);
+        expect(spy.callCount).to.equal(3);
```

Comment:
this line (or the title of this test) should have a comment saying that the fact that `callCount` is `3` asserts that `setState` inside `componentWillReceiveProps` *does* provoke another render.